### ISSUE TITLE
Where to put helper files for extensions?

### ISSUE THREAD

**Issue by JackLeEmmerdeur** [directus/docs#176]
# Bug Report

Trying to include a PHP file with helper functions within the hooks.php 

## Steps to Reproduce

1. Create the hooks.php file in public/extensions/hooks
2. Create another PHP file with helper functions within the same path
3. require("helpers.php") within hooks.php and use a function there

## Expected Behavior
The function should execute in hooks.php

## Actual Behavior
When I create the helper.php file in step 2, the login page of directus becomes dysfunctional and the browser-console throws "GET http://directus7.dev.local/_/auth/sso 500 (Internal Server Error)". The helpers.php file has the same file-permissions and owner/group as hooks.php.
Where within then extensions folder should I create the afore mentioned helper-file?

## Technical Details

- Device: Desktop
- OS: Debian 9
- Web Server: nginx/1.10.3
- PHP Version: PHP-FPM 7.3.7
- Database: MariaDB 10.3.16
- Install Method: cloned `master` branch

**Comment by benhaynes** [comment]
Thanks @JackLeEmmerdeur — I'll move this to Docs so we remember to add some info on this!

**Comment by JackLeEmmerdeur** [comment]
That could prevent some headscratching @benhaynes. Didn't know the other repositories (docs, api etc.). Another place to search issues if I'm lost. If any help with the documentation is needed, I could do on and off tasks (e.g. this issue or if planned translations to german).

**Comment by benhaynes** [comment]
That would be awesome! All of our Docs are open-source repos of Markdown — so feel free to make any suggestions/edits/additions that you think might help clear things up. You can do them directly from GitHub, so it should be pretty easy. :)

For example...
https://github.com/directus/docs/edit/master/extensions/README.md
Turns into...
https://docs.directus.io/extensions/

**Comment by benhaynes** [comment]
Our Admin App is in German... but the Docs are only English for now. Once we have everything stable and documented we'll be looking into multilingual docs. :)

**Comment by JackLeEmmerdeur** [comment]
I previously submitted only one pull-request (accepted). I guess the process is the same here, isn't it? Fork > Checkout new branch >Edit > Commit > Push to Fork > Create Pull Request. If I completely misunderstood are there any docs for reference?

**Comment by benhaynes** [comment]
Yeah, that's basically it! You can do it the simple (GitHub) or normal (locally) way:

https://docs.directus.io/getting-started/contributing.html#simple-pull-requests
https://docs.directus.io/getting-started/contributing.html#proper-pull-requests

Since the Docs are just markdown you can probably get away with the simple way if that's easier!

**Comment by JackLeEmmerdeur** [comment]
Thanks for that heads-up. Clearly a symptom of RTFM on my side not spotting [this](https://docs.directus.io/getting-started/contributing.html). I'll try to add the topic to the hooks page, asap (today or tomorrow) possibly within a ::: warning block. Is the idea of putting an additional folder within the hooks-folder for own php-files an appropiate way, to propose in the docs?

**Comment by rijkvanzanten** [comment]
@JackLeEmmerdeur It's the same way for endpoints (which _is_ documented... a bit): 

https://docs.directus.io/extensions/custom-endpoints.html#using-directories

so I'd say it's perfectly fine 🙂

**Comment by JackLeEmmerdeur** [comment]
@rijkvanzanten Again a case of RTFM. I'm glad you pointed me to it. You can write own APIs then. Nice.

I guess by "a bit" you pointed to "You can use any file-names or sub-directories to help keep things organized.".

So the root-directory-layout of hooks and endpoints are similar/same? Just found out by experimenting that you could create multiple files (e.g. producthooks.php and pricehooks.php) in the   extensions/custon/hooks folder. Essential is that they return an array with filter or action definitions.

One thing I would like to understand is how the subfolders work. To activate a extension/hook/endpoint you remove the leading underscore of the folder. But if I do that in my environment with the provided example "_products", I get the afore mentioned 500 status and a spinning progressbar on the login-screen. No suspicious entrys in php, nginx, mysql or the logs/error.2019-07-25.log. 

I want that sorted out before I write about a thing I've never seen in action. Going to dig in the issues, for now.

**Comment by JackLeEmmerdeur** [comment]
No luck with sorting out the HookInterface implementing-class in a subfolder. Be it:
 
- hooks.php and the class in subfolder
- both in the root of the hooks-folder
- hooks.php in hooks-folder and class in subfolder
- different ways to define the filter-array-value:
   'filter' => new \Directus\Custom\Hooks\Products\BeforeInsertProducts()
   'filter' =>\Directus\Custom\Hooks\Products\BeforeInsertProducts::class
   'filter' =>"\Directus\Custom\Hooks\Products\BeforeInsertProducts"
   All with and without leading backslash
- the configuration of filters in api.php,
- instead of implementing HookInterface using the __invoke method in a POC

So I'm back to defining a hooks/ProductHooks.php and in it the simple array with closure-functions for the filter.

At least I learned that you can organize your hooks by creating multiple files in the hooks-folder and that I can require helper-files from subfolders.

I'll try to transport that to the docs, but will spare out the HookInterface.

**Comment by rijkvanzanten** [comment]
> I guess by "a bit" you pointed to "You can use any file-names or sub-directories to help keep things organized.".

What I meant is that it's documented _a bit_, but not documented that well.

**Comment by JackLeEmmerdeur** [comment]
It appears that the problems I had were not only a result of RTFM, but of PEBKAC, too. Seems that I messed up at installation already.

Had been running the previous version, with near 0 errors. 

Then did a git-clone install for a new project, but the API installation site did not appear when hitting index.php for the first time. Maybe because a directus-database already existed. Ended up with installing [manually](https://docs.directus.io/advanced/api/configuration.html) to a differently named database.

Since working with the latest realease, I thought there were some serious regressions.

Luckily I've installed again via git-clone. All issues are gone. I have not validated that my problems with the HookInterface are gone, but I'll test when I've got the collections setup. I'm pretty confident it'll work. If not I'll open another issue.

Nevertheless I already forked docs and added stuff to the hooks.md, regarding file/folder structures. Hope to create a pull-request the next days.

### **TL;DR**
I'm still working on the docs, but if anyone is quicker, first come first served.
As I deviated from the topic a little, could this be closed?
